File: src/main/java/com/airsaid/localization/translate/impl/googleapi/GoogleApiTranslationResult.java
package com.airsaid.localization.translate.impl.googleapi;

import com.airsaid.localization.translate.TranslationResult;
import org.jetbrains.annotations.NotNull;

import java.util.Arrays;

/**
 * @author airsaid
 */
public class GoogleApiTranslationResult implements TranslationResult {
  public Data data;
  public Error error;

  public boolean isSuccess() {
    return data != null && error == null;
  }

  @Override
  public @NotNull String getTranslationResult() {
    return data.translations[0].translatedText;
  }

  @Override
  public String toString() {
    return "GoogleApiTranslationResult{" +
        "data=" + data +
        ", error=" + error +
        '}';
  }

  public static class Data {
    public Translation[] translations;

    @Override
    public String toString() {
      return "Data{" +
          "translations=" + Arrays.toString(translations) +
          '}';
    }

    public static class Translation {
      public String translatedText;
      public String detectedSourceLanguage;

      @Override
      public String toString() {
        return "Translation{" +
            "translatedText='" + translatedText + '\'' +
            ", detectedSourceLanguage='" + detectedSourceLanguage + '\'' +
            '}';
      }
    }
  }

  public static class Error {
    public int code;
    public String message;

    @Override
    public String toString() {
      return "Error{" +
          "code=" + code +
          ", message='" + message + '\'' +
          '}';
    }
  }
}

File: src/main/kotlin/com/airsaid/localization/translate/impl/googleapi/GoogleApiTranslationResult.kt
package com.airsaid.localization.translate.impl.googleapi

import com.airsaid.localization.translate.TranslationResult

/**
 * @author airsaid
 */
data class GoogleApiTranslationResult(
    var data: Data? = null,
    var error: Error? = null
) : TranslationResult {

    fun isSuccess(): Boolean = data != null && error == null

    override val translationResult: String
        get() = data?.translations?.get(0)?.translatedText ?: ""

    data class Data(
        var translations: Array<Translation>? = null
    ) {
        data class Translation(
            var translatedText: String? = null,
            var detectedSourceLanguage: String? = null
        )

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (javaClass != other?.javaClass) return false

            other as Data

            if (translations != null) {
                if (other.translations == null) return false
                if (!translations.contentEquals(other.translations)) return false
            } else if (other.translations != null) return false

            return true
        }

        override fun hashCode(): Int {
            return translations?.contentHashCode() ?: 0
        }
    }

    data class Error(
        var code: Int = 0,
        var message: String? = null
    )
}